### PR TITLE
queryparser: added safelisting of terms

### PR DESCRIPTION
**Pull request by alejandromumo**
closes https://github.com/inveniosoftware/invenio-records-resources/issues/447

### CODE REVIEW

File: invenio_records_resources/services/records/queryparser/transformer.py [hunk 1]
```diff
@@ -24,24 +24,30 @@
 class SearchFieldTransformer(TreeTransformer):
     """Transform from user-friendly field names to internal field names."""
 
-    def __init__(self, mapping, *args, **kwargs):
+    def __init__(self, mapping, safelist, *args, **kwargs):
```
Comment (alejandromumo) [inline invenio_records_resources/services/records/queryparser/transformer.py:27]: This can also be implemented in the `QueryParser` as a "step" of the parsing process. 
I added it here for now since we're visiting one field at the time (using `luqum`). So it was easier to do it here and didn't feel completely wrong.

But please challenge it :)

**Comment by alejandromumo** [comment]
I can fix tests when this is accepted, since now it's a "WIP" I will not fix them and focus on other tasks.

File: invenio_records_resources/services/records/queryparser/transformer.py [hunk 2]
```diff
@@ -24,24 +24,30 @@
...
+        return partial(cls, mapping or {}, safelist or [])
 
     def visit_search_field(self, node, context):
         """Visit a search field."""
-        if node.name not in self._mapping:
+        # Use the node name if not mapped for transformation.
+        term_name = self._mapping.get(node.name, node.name)
+
+        # If a safelist exists, the term must be safelisted to be queried.
+        if self._safelist and not term_name in self._safelist:
```
Comment (ppanero) [inline invenio_records_resources/services/records/queryparser/transformer.py:44]: Do we have a case where we would not want a safelist? records? This is to make our lives easier when defining the parsing?

Comment (alejandromumo) [reply invenio_records_resources/services/records/queryparser/transformer.py:44]: Sorry for the late reply!

Yes there are searches that we do not want / need a safelist. Initially the terms mapping also acted as a safelist and that was particularly useful for users becasuse some terms are not meant to be queried.

However, since we shipped two features in one (mapping + safelisting) it also meant that if a transform map was defined, then any other terms are considered were blocked (e.g. can't be queried at all). In fact, this had a side effect on which the transformation map needed things like `{"email": "email"}` since the term was not transformed AND being blocked if not defined.

TL;DR : yes we do have cases where we do not want to have a safelist. And yes, this separation of features makes our lives easier.

**Review by ppanero — APPROVED** [review]

**Review by ppanero — APPROVED** [review]
I have mostly questions for my understanding, I don't think any should block the merge.

File: tests/services/test_service_queryparser.py [hunk 3]
```diff
@@ -96,21 +96,23 @@ def test_multimatch(parser, query):
...
-        ("description:test", "description:test", "multi_match"),
-        ("metadata.title:test", "metadata.title:test", "multi_match"),
+        ("title:test", "metadata.title:test", "query_string", None),
+        ("title:(test test)", "metadata.title:(test test)", "query_string", None),
+        ("title:[1 TO 5]", "metadata.title:[1 TO 5]", "query_string", None),
+        # Term is allowed and not mapped, results in a query string with the original query
+        ("description:test", "description:test", "query_string", None),
+        ("metadata.title:test", "metadata.title:test", "query_string", None),
+        # Terms might be mapped but not allowed, results in multi_match with the original query
+        ("title:test", "title:test", "multi_match", ["metadata.date"]),
```
Comment (ppanero) [inline tests/services/test_service_queryparser.py:108]: I'm missing something (because tests pass). Why isn't this entering `transformer.py#44`? The query term `title` is in the mapping and there is an allowed list without it so:

```
if self._allow_list and not term_name in self._allow_list:

if ["metadata.date"] and not `title` in ["metadata.date"]:

True and not False:
```

I'm puzzled 😂

Comment (ppanero) [reply tests/services/test_service_queryparser.py:108]: https://github.com/inveniosoftware/invenio-records-resources/blob/0547e5d7cb7d30cca5c470e6fb264de408ce77ed/invenio_records_resources/services/records/queryparser/query.py#L92

**Review by ppanero — COMMENTED** [review]
Discussed with Nico and Karolina. The two query parser issues will be tackled independenty here https://github.com/inveniosoftware/invenio-records-resources/issues/464.

Therefore, there is only a need for updating (ref [search](https://github.com/search?q=org%3Ainveniosoftware%20QueryParser&type=code))
- https://github.com/inveniosoftware/docs-invenio-rdm/blob/63923ea359ad4ab41bc37440098d59d0fe494d26/docs/develop/howtos/search_terms_migration.md?plain=1#L23
- https://github.com/inveniosoftware/invenio-users-resources/blob/10d739c945e371366026d6f0156566c87cfb5f1f/invenio_users_resources/services/groups/config.py#L38 <- Not sure about this one

**Comment by ppanero** [comment]
Needs updating (ref [search](https://github.com/search?q=org%3Ainveniosoftware%20QueryParser&type=code))
- https://github.com/inveniosoftware/docs-invenio-rdm/blob/63923ea359ad4ab41bc37440098d59d0fe494d26/docs/develop/howtos/search_terms_migration.md?plain=1#L23
- https://github.com/inveniosoftware/invenio-users-resources/blob/10d739c945e371366026d6f0156566c87cfb5f1f/invenio_users_resources/services/groups/config.py#L38 <- Not sure about this one

File: invenio_records_resources/services/records/queryparser/transformer.py [hunk 4]
```diff
@@ -24,24 +24,30 @@
...
                 _("Invalid search field: {field_name}.").format(field_name=node.name)
             )
-        else:
-            new_node = node.clone_item()
-            new_node.name = self._mapping[node.name]
-            new_node.children = list(self.clone_children(node, new_node, context))
-            yield new_node
+
+        # Returns a copy of the node.
+        new_node = node.clone_item()
```
[the remark this answers is not in the document]
Comment (lnielsen) [reply invenio_records_resources/services/records/queryparser/transformer.py:50]: I think it comes from the [luqum docs example](https://luqum.readthedocs.io/en/latest/quick_start.html#manipulating) and that it's the intended way to do it: https://luqum.readthedocs.io/en/latest/api.html#luqum.tree.Item.clone_item
